ta/pkcs11/src/sanitize_object.c

```diff
@@ -0,0 +1,428 @@
+			return rc;
+		}
+
+		rc = add_attribute(dst, cli_ref.id, data, cli_ref.size);
+		if (rc)
+			return rc;
+	}
+
+	return rc;
+}
+
+/*
+ * Debug: dump object attribute array to output trace
+ */
+
+static void __trace_attributes(char *prefix, void *src, void *end)
+{
+	size_t next = 0;
+	char *prefix2 = NULL;
+	size_t prefix_len = strlen(prefix);
+	char *cur = src;
+
+	/* append 4 spaces to the prefix plus terminal '\0' */
+	prefix2 = TEE_Malloc(prefix_len + 1 + 4, TEE_MALLOC_FILL_ZERO);
+	if (!prefix2)
+		return;
+
+	TEE_MemMove(prefix2, prefix, prefix_len + 1);
+	TEE_MemFill(prefix2 + prefix_len, ' ', 4);
+	*(prefix2 + prefix_len + 4) = '\0';
+
+	for (; cur < (char *)end; cur += next) {
+		struct pkcs11_attribute_head pkcs11_ref;
+		uint8_t data[4] = { 0 };
+		uint32_t data_u32 = 0;
+
+		TEE_MemMove(&pkcs11_ref, cur, sizeof(pkcs11_ref));
+		TEE_MemMove(&data[0], cur + sizeof(pkcs11_ref),
+			    MIN(pkcs11_ref.size, sizeof(data)));
+		TEE_MemMove(&data_u32, cur + sizeof(pkcs11_ref),
+			    sizeof(data_u32));
+
+		next = sizeof(pkcs11_ref) + pkcs11_ref.size;
+
+		IMSG_RAW("%s Attr %s / %s (%#04"PRIx32" %"PRIu32"-byte)",
+			 prefix, id2str_attr(pkcs11_ref.id),
+			 id2str_attr_value(pkcs11_ref.id, pkcs11_ref.size,
+					   cur + sizeof(pkcs11_ref)),
+			 pkcs11_ref.id, pkcs11_ref.size);
+
+		switch (pkcs11_ref.size) {
+		case 0:
+			break;
+		case 1:
+			DMSG_RAW("%s Attr byte value: %02x", prefix, data[0]);
+			break;
+		case 2:
+			DMSG_RAW("%s Attr byte value: %02x %02x",
+				 prefix, data[0], data[1]);
+			break;
+		case 3:
+			DMSG_RAW("%s Attr byte value: %02x %02x %02x",
+				 prefix, data[0], data[1], data[2]);
+			break;
+		case 4:
+			DMSG_RAW("%s Attr byte value: %02x %02x %02x %02x",
+				 prefix, data[0], data[1], data[2], data[3]);
+			break;
+		default:
+			DMSG_RAW("%s Attr byte value: %02x %02x %02x %02x ...",
+				 prefix, data[0], data[1], data[2], data[3]);
+			break;
+		}
+
+		switch (pkcs11_ref.id) {
+		case PKCS11_CKA_WRAP_TEMPLATE:
+		case PKCS11_CKA_UNWRAP_TEMPLATE:
+		case PKCS11_CKA_DERIVE_TEMPLATE:
+			trace_attributes_from_api_head(prefix2,
+						       cur + sizeof(pkcs11_ref),
+						       (char *)end - cur);
+			break;
+		default:
+			break;
+		}
+	}
+
+	/* Sanity */
+	if (cur != (char *)end)
+		EMSG("Warning: unexpected alignment issue");
+
+	TEE_Free(prefix2);
+}
+
+void trace_attributes_from_api_head(const char *prefix, void *ref, size_t size)
+{
+	struct pkcs11_object_head head = { };
+	char *pre = NULL;
+	size_t offset = 0;
+
+	TEE_MemMove(&head, ref, sizeof(head));
+
+	if (size > sizeof(head) + head.attrs_size) {
+		EMSG("template overflows client buffer (%zu/%zu)",
+		     size, sizeof(head) + head.attrs_size);
+		return;
+	}
+
+	pre = TEE_Malloc(prefix ? strlen(prefix) + 2 : 2, TEE_MALLOC_FILL_ZERO);
+	if (!pre) {
+		EMSG("%s: out of memory", prefix);
+		return;
+	}
+	if (prefix)
+		TEE_MemMove(pre, prefix, strlen(prefix));
+
+	IMSG_RAW("%s,--- (serial object) Attributes list --------", pre);
```

Comment:
`DMSG_RAW()` since `DMSG` used in `__trace_attributes()`.

Reply:
OK, updating.